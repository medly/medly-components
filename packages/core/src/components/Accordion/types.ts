import { WithThemeProp } from '@medly-components/utils';

export interface Props extends WithThemeProp{
    /** Title of the accordion */
    title: string;
    /** Color for the Accordion title*/
    titleColor?: string;
    /** Background color for the Accordion section*/
    bgColor?: string;

}

export interface StyleProps extends Omit<Props, 'title' | 'content'> {
    isActive?: boolean;
}


