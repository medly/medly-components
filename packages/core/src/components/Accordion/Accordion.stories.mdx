import Text from '../Text';
import { defaultTheme } from '@medly-components/theme';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import { withKnobs, color, select } from '@storybook/addon-knobs';
import { SimpleAccordion, ThemeInterface } from './Accordion.stories';
import { Accordion } from './Accordion';


<Meta title="Core" component={Accordion} parameters={{ jest: ['Accordion.test.tsx'] }} />

# Accordion

Accordion is a container that can either stand alone or be connected with other components.

<Preview withToolbar>
    <Story name="Accordion" parameters={{ decorators: [withKnobs] }}>
        <Accordion
            title="Uploaded Files"
            bgColor={color('Background color', defaultTheme.accordion.backgroundColor)}
            titleColor={color('Title Color', defaultTheme.colors.black)}
        >
            <div>File 1</div>
        </Accordion>
    </Story>
</Preview>


### Props

<Props of={Accordion} />

#### Accordion Theme

<Props of={ThemeInterface} />
