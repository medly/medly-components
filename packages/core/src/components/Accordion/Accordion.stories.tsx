import { AccordionTheme, defaultTheme } from '@medly-components/theme';
import React from 'react';

export const ThemeInterface: React.SFC<AccordionTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.accordion
};
