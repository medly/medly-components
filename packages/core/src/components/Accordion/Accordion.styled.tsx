import { SvgIcon } from '@medly-components/icons';
import styled from 'styled-components';
import { StyleProps } from './types';

export const Section = styled('div').attrs(({ theme: { accordion: { backgroundColor } } }: StyleProps) => ({
    backgroundColor
}))<StyleProps>`
    display: flex;
    flex-direction: column;
    background-color: ${({ bgColor, backgroundColor }) => bgColor || backgroundColor};
    border-radius: 0.8rem;
`;

export const Wrapper = styled('button')<StyleProps>`
    cursor: pointer;
    padding: 1.8rem;
    display: flex;
    align-items: center;
    outline: none;
    border: none;
    background-color: transparent;

    ${SvgIcon} {
        margin-left: auto;
        transition: transform 0.25s ease;
        transform: ${({ isActive }) => (isActive ? 'rotate(180deg)' : 'rotate(0deg)')};
    }
`;

export const Content = styled('div').attrs(({ theme: { accordion: { separatorColor } } }: StyleProps) => ({
    separatorColor
}))<StyleProps>`
    display: flex;
    transition: width 2s;
    transition-timing-function: linear;
    padding: 1.8rem;
    border-top: solid 0.2rem ${({ separatorColor }) => separatorColor};
`;
