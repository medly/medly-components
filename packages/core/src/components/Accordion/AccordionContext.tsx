import { createContext } from 'react';
import { AccordionContextType } from './types';

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const AccordionContext = createContext<AccordionContextType>([false, () => {}]);
