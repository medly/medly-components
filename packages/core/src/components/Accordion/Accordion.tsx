import { ExpandMoreIcon } from '@medly-components/icons';
import React, { SFC, useCallback, useState } from 'react';
import Text from '../Text';
import { Content, Section, Wrapper } from './Accordion.styled';
import { Props } from './types';

export const Accordion: SFC<Props> = React.memo(({ title, titleColor, bgColor, children }) => {
    const [active, setActive] = useState(false),
        toggleAccordion = useCallback(() => setActive(val => !val), []);

    return (
        <Section bgColor={bgColor}>
            <Wrapper isActive={active} onClick={toggleAccordion}>
                <Text textVariant="body1" textColor={titleColor}>
                    {title}
                </Text>
                <ExpandMoreIcon />
            </Wrapper>

            {active && <Content>{children}</Content>}
        </Section>
    );
});

Accordion.displayName = 'Accordion';
