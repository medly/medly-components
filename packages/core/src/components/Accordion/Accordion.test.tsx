import { render } from '@test-utils';
import React from 'react';
import { Accordion } from './Accordion';
import { fireEvent, waitFor } from '@testing-library/dom';

describe('Accordion component', () => {

    const renderAccordion = () =>
        render(
            <Accordion title="Test Accordion">
                <p>Test Text</p>
            </Accordion>
        )
    it('should render accordion with default prop values', () => {
        const { container } = renderAccordion();
        expect(container).toMatchSnapshot();
    });

    it('should render children when accordion is clicked', async () => {
        const { getByText, queryByText  } = renderAccordion();
        fireEvent.click(getByText(/Test Accordion/));
        await waitFor(() => expect(getByText(/Test Text/)).toBeInTheDocument());
        fireEvent.click(getByText(/Test Accordion/));
        expect(queryByText(/Test Text/)).not.toBeInTheDocument();
    });
});
