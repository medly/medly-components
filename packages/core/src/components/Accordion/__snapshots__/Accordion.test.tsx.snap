// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component should render accordion with default prop values 1`] = `
.c4 {
  font-size: 2.4rem;
  cursor: inherit;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
}

.c4 * {
  fill-opacity: 1;
  fill: #607890;
}

.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.6rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.6rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #ffffff;
  border-radius: 0.8rem;
}

.c1 {
  cursor: pointer;
  padding: 1.8rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  outline: none;
  border: none;
  background-color: transparent;
}

.c1 .c3 {
  margin-left: auto;
  -webkit-transition: -webkit-transform 0.25s ease;
  -webkit-transition: transform 0.25s ease;
  transition: transform 0.25s ease;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1"
    >
      <span
        class="c2"
      >
        Test Accordion
      </span>
      <svg
        class="c3 c4"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.875 9l-3.88 3.88L8.115 9a.996.996 0 1 0-1.41 1.41l4.59 4.59c.39.39 1.02.39 1.41 0l4.59-4.59a.996.996 0 0 0 0-1.41c-.39-.38-1.03-.39-1.42 0z"
          fill="#000"
          fill-opacity=".54"
        />
      </svg>
    </button>
  </div>
</div>
`;
